extensions/git/src/knownFolders.ts

```diff
@@ -0,0 +1,143 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Memento, workspace } from 'vscode';
+import * as path from 'path';
+import { LRUCache } from './cache';
+
+export class KnownFolders {
+
+	private static readonly STORAGE_KEY = 'git.knownFolders';
+	private static readonly MAX_REPO_ENTRIES = 30; // Max repositories tracked
+	private static readonly MAX_FOLDER_ENTRIES = 10; // Max folders per repository
+	private static readonly EXPIRY_MS = 90 * 24 * 60 * 60 * 1000; // 90 days in ms
+
+	// Outer LRU: repoUrl -> inner LRU (folderPathOrWorkspaceFile -> timestamp ms). Only keys matter externally, timestamp tracks last seen.
+	private readonly lru = new LRUCache<string, LRUCache<string, number>>(KnownFolders.MAX_REPO_ENTRIES);
+
+	constructor(public readonly _globalState: Memento) {
+		this.load();
+	}
+
+	/**
+	 * Associate a repository remote URL with a local workspace folder or workspace file.
+	 * Re-associating bumps recency and persists the updated LRU state.
+	 * @param repoUrl Remote repository URL (e.g. https://github.com/owner/repo.git)
+	 * @param rootPath Root path of the local repo clone.
+	 */
+	set(repoUrl: string, rootPath: string): void {
+		let foldersLru = this.lru.get(repoUrl);
+		if (!foldersLru) {
+			foldersLru = new LRUCache<string, number>(KnownFolders.MAX_FOLDER_ENTRIES);
+		}
+		// If the current workspace is a workspace file, use that. Otherwise, find the workspace folder that contains the rootUri
+		let folderPathOrWorkspaceFile: string | undefined;
+		try {
+
+			if (workspace.workspaceFile) {
+				folderPathOrWorkspaceFile = workspace.workspaceFile.fsPath;
+			} else if (workspace.workspaceFolders && workspace.workspaceFolders.length) {
+				const sorted = [...workspace.workspaceFolders].sort((a, b) => b.uri.fsPath.length - a.uri.fsPath.length);
+				for (const folder of sorted) {
+					const folderPath = folder.uri.fsPath;
+					const rel = path.relative(folderPath, rootPath);
+					if (rel === '' || (!rel.startsWith('..') && !path.isAbsolute(rel))) {
+						folderPathOrWorkspaceFile = folderPath;
+						break;
+					}
+				}
+			}
+
+			if (!folderPathOrWorkspaceFile) {
+				folderPathOrWorkspaceFile = rootPath;
+			}
+		} catch {
+			return;
+		}
+
+
+		foldersLru.set(folderPathOrWorkspaceFile, Date.now()); // touch/update timestamp
+		this.lru.set(repoUrl, foldersLru);
+		this.save();
+	}
+
+	/**
+	 * We should possibly support converting between ssh remotes and http remotes.
```

Comment:
I would probably "normalize" the remote when storing it but we can tackle this later.